app/forms/otp_verifications_form.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+class OtpVerificationsForm
+  include ActiveModel::Model
+  include ActionView::Helpers::DateHelper
+
+  validate :cool_down?
+  validates :code, presence: true
+  validate :code_is_correct?
+
+  def initialize(session:, code:)
+    @session = session
+    @code = code
+
+    @user = User.find_by(email: session[:otp_email])
+    @salt = session[:otp_salt]
+    @attempts = session[:otp_attempts] = session[:otp_attempts] || 0
+    @last_attempt = session[:otp_last_attempt] = session[:otp_last_attempt] || Time.zone.now
+  end
+
+private
+
+  attr_reader :session, :code, :user, :salt, :attempts, :last_attempt
+
+  # 600 = 10 mins validity
+  def code_is_correct?
+    unless user && totp.verify(code, drift_behind: 600).present?
+      errors.add(:code, :invalid_code)
+    end
+  end
+
+  def cool_down?
+    # Add error message if the user has not yet reached their
+    # cool off time.
+    if Time.zone.now < (last_attempt + cool_down_time)
+      errors.add(:code, cool_down_message)
+    # Otherwise incement their attempts and reset the last attempt
+    # if they submitted a code
+    elsif code.present?
+      session[:otp_attempts] += 1
+      session[:otp_last_attempt] = Time.zone.now
+    end
+  end
+
+  def cool_down_message
+    @cool_down_message ||= "Please wait #{time_left} before trying again"
+  end
+
+  # increase cool off then default to 1 hour
+  def cool_down_time
+    @cool_down_time ||= ([0, 0, 0, 0, 0, 60, 300, 600][attempts] || 3600).seconds
```

Comment:
@asatwal yes, that's the intent. I came up with these numbers. Aiming that you can make several initial attempts, but after that we slow you down. One of the benefits of this is not needing a hard account lock system and processes for 'unlocking' an account if someone does go over.

It would give ~30 attempts a day. Though the expiry of the code is currently 10 minutes - though someone could get a new code. It's intended these limits persist across new codes until you get it correct.

I wonder if we should have any security logging on this if we get lots of attempts?